lib/zk_recipes/cache.rb

```diff
@@ -205,31 +508,148 @@ def process_pending_updates
+        @logger&.debug { "process_pending_updates: path=#{path} type=#{type}" }
+        delete = case type
+                 when :static
+                   update_static_path(path)
+                 when :runtime
+                   update_runtime_path(path)
+                 when :directory
+                   update_directory_path(path)
+                 else
+                   @logger&.error { "process_pending_updates: unknown pending_update type=#{type.inspect}" }
+                   true
+                 end
+        @pending_updates.delete(path) if delete
+      end
+
+      if @pending_updates.empty?
+        @logger&.debug("process_pending_updates: processed all pending updates")
+      else
+        @logger&.warn { "process_pending_updates: failed to process pending updates=#{@pending_updates.inspect}" }
       end
     end
 
-    class CachedPath
-      attr_reader :value, :stat
-      def initialize(value, stat: nil, valid: false)
-        @value = value
+    class StaticPath
+      def initialize(default_value, deserializer)
+        @mutex = Mutex.new
+        @value = @default_value = default_value
+        @deserializer = deserializer
+        @stat = nil
+        @valid = false
+      end
+
+      def value
+        @mutex.synchronize { @value }
+      end
+
+      def valid_value
+        @mutex.synchronize { @value if @valid }
+      end
+
+      def reset_value!(stat)
+        @mutex.synchronize do
+          @value = @default_value
+          @valid = false
+        end
         @stat = stat
-        @valid = valid
       end
 
-      def valid?
-        @valid
+      def deserialize(raw, stat = nil)
+        @stat = stat if stat
+        value = @deserializer ? @deserializer.call(raw) : raw
+        valid = true
+        if value == USE_DEFAULT
+          value = @default_value
+          valid = false
+        end
+        @mutex.synchronize do
+          @valid = valid
+          @value = value
+          @value
+        end
+      rescue
+        @mutex.synchronize do
+          @value = @default_value
+          @valid = false
+        end
+        raise
       end
     end
-    private_constant :CachedPath
+    private_constant :StaticPath
+
+    class Directory
+      def initialize(path_mapper, deserializer)
+        @mutex = Mutex.new
+        @path_mapper = path_mapper
+        @deserializer = deserializer
+
+        @watched_paths = {}
+        @paths_with_values = {}
+        @stat = nil
+      end
+
+      def reset_paths!(stat)
+        paths = @watched_paths.keys
+        @mutex.synchronize do
+          @watched_paths.clear
+          @paths_with_values.clear
+        end
+        @stat = stat
+        paths
+      end
+
+      def map_paths(paths, stat)
+        @stat = stat
+        watched_paths = {}
+        new_paths = {}
+        removed_paths = nil
+        paths = paths.map { |path| @path_mapper.call(path) }
+        @mutex.synchronize do
+          removed_paths = @watched_paths
```

Comment:
It seems like `@watched_paths` is really just a set? Any particular reason to not use a `Set`?

With sets, I think you could simplify this to: 
```ruby
def map_paths(paths, stat)
  @stat = stat
  removed_paths = nil
  added_paths = nil
  incoming_paths = Set.new(paths.map { |path| @path_mapper.call(path) })

  @mutex.synchronize do
    removed_paths = @watched_paths.difference(incoming_paths)
    added_paths = incoming_paths.difference(@watched_paths)

    @watched_paths = incoming_paths
  end
  [added_paths, removed_paths]
end
```

Reply:
It is just a set. I had one little issue with the set API and changed to hashes. This does seem to simplify it a bit. I'll look at switching it back to a set.